src/scene/particle-container/shared/ParticleContainer.ts

```diff
@@ -0,0 +1,571 @@
+     * @returns - The particle at the given index, if any.
+     */
+    public getParticleAt<U extends IParticle>(index: number): U
+    {
+        if (index < 0 || index >= this.particleChildren.length)
+        {
+            throw new Error(`getChildAt: Index (${index}) does not exist.`);
+        }
+
+        return this.particleChildren[index] as U;
+    }
+
+    /**
+     * Changes the position of an existing particle in the container
+     * @param child - The particle instance for which you want to change the index number
+     * @param index - The resulting index number for the particle
+     */
+    public setParticleIndex(child: IParticle, index: number)
+    {
+        const _index = this.particleChildren.indexOf(child);
+
+        this.particleChildren.splice(_index, 1);
+        this.particleChildren.splice(index, 0, child);
+
+        this.onViewUpdate();
+
+        return child;
+    }
+
+    /**
+     * Returns the index position of a particle instance
+     * @param child - The particle instance to identify
+     * @returns - The index position of the particle to identify
+     */
+    public getParticleIndex(child: IParticle)
+    {
+        return this.particleChildren.indexOf(child);
+    }
+
+    /**
+     * Adds a particle to the container at a specified index. If the index is out of bounds an error will be thrown.
+     * If the particle is already in this container, it will be moved to the specified index.
+     * @param {Container} child - The particle to add.
+     * @param {number} index - The absolute index where the particle will be positioned at the end of the operation.
+     * @returns {Container} The particle that was added.
+     */
+    public addParticleAt<U extends IParticle>(child: U, index: number): U
+    {
+        this.particleChildren.splice(index, 0, child);
+
+        this.onViewUpdate();
+
+        return child;
+    }
+
+    /**
+     * Swaps the position of 2 particles within this container.
+     * @param child - First container to swap
+     * @param child2 - Second container to swap
+     */
+    public swapParticles(child: IParticle, child2: IParticle)
+    {
+        const index = this.particleChildren.indexOf(child);
+        const index2 = this.particleChildren.indexOf(child2);
+
+        this.particleChildren[index] = child2;
+        this.particleChildren[index2] = child;
+
+        this.onViewUpdate();
+
+        return child;
+    }
+
+    /**
+     * This method is not available in ParticleContainer.
+     *
+     * Calling this method will throw an error. Please use `ParticleContainer.addParticle()` instead.
+     * @param {...any} _children
+     * @throws {Error} Always throws an error as this method is not available.
+     */
+    public override addChild<U extends ContainerChild[]>(..._children: U): U[0]
+    {
+        throw new Error(
+            'ParticleContainer.addChild() is not available. Please use ParticleContainer.addParticle()',
+        );
+    }
```

Comment:
@bigtimebuddy @GoodBoyDigital 

For `ParticleContainer` I have found that you can't accurately convert the types for `addChild`/`removeChild` etc to return `IParticle` without losing type accuracy for normal containers.

So the current solution is to change the names of these helper function to `addParticle`/ `removeParticle` etc
and then to throw errors if a user tries to use one of the "normal" functions

It's not quite as clean as I would have liked but it does keep type accuracy, any thoughts on this approach?

also as an additional question, should we support the less common helpers such as `swapChildren`/`getChildIndex` etc, or just keep to `addParticle`/`removeParticle`/`addParticleAt`/`removeParticleAt`/`removeParticles`

Reply:
I think this is a nice solution to not only the type issue but also what might be developer confusion between Container and ParticleContainer. Having a harder API break is helpful, I think.


Might be nice to also have some development errors thrown if the child is not an instanceof Particle. I can definitely see v7 users trying to add Sprites here.

Reply:
Be conservative about all the container methods. We can always easily add later if folks use this heavily.